### ISSUE TITLE
Naturel landcover script

### ISSUE THREAD

**Issue by matthiasdemuzere** [matthiasdemuzere/wrf-lcz-KL#13]
Hi @dargueso,

I am cleaning up my script a bit, and it would be nice if I could convert the wrf_remove_urban function, which I took from your original script, into something that uses xarray instead of netCDF4. That would make everything more consistent.

Are you familiar with xarray? If so, could you try to convert this routine?
If not, I can try myself, although I am not sure what exactly is happening in this routine ...

**Comment by matthiasdemuzere** [comment]
> I've pushed changes to the wrf_remove_urban function. It now uses xarray. It's not the cleanest code.
> It includes a few changes that I think are necessary for the tool.

This is great, thanks
> 
> 1. Greenfrac is calculated using mean and not mode. Mode makes sense when using integers, not so much when using floats.

Ok. So what exactly does a greenfrac value for a pixel now mean? Is is the average greenfrac value over all pixels that were sampled from a natural class being the majority in your 45pixel buffer? Or just the mean over all natural pixels in that buffer?

> 2. The way landusef is replaced ensures the sum is one. It takes the original value of landusef for urban category, and assigns that to the new lu_index. It also makes more sense, since if a grid cell has 0.5 of urban, that 0.5 is replaced with the new landuse. The remaining 0.5 remains untouched.

Great.

> 3. When replacing with the surrounding dominant landuse, we don't include lakes. This wasn't a problem in our domain, but it may be in different locations.

Ok, so this is fixed now? Or how do you take this into account?

> 
> Our NoUrban file may not fulfil all these conditions, and the resulting file is slightly different. But at this point, the NoUrban simulations is almost complete. I thought that, since we're going to share the tool, it should be as rigorous as possible.

I agree. I don't think it matters too much for our experiment?

**Comment by matthiasdemuzere** [comment]
I just re-formatted your code a bit (avoiding long lines), and will try this out. If it works, I'll push these small changes back.

**Comment by dargueso** [comment]
Yes! You can find them in @matthiasdemuzere [GD ](https://drive.google.com/drive/folders/15QsWRCxFEjZ7fo3C6iP9vqISPr5uw6l1).

**Comment by andreazonato** [comment]
That's great! Everything is fine! You should be able to run the
simulation...

Il giorno mar 11 mag 2021 alle ore 17:37 Daniel Argüeso <
***@***.***> ha scritto:

**Comment by dargueso** [comment]
I'm struggling to get exp C running. WRF is crashing without giving any error, which is not surprising in WRF anyway. I'm trying to figure out what is causing the error with high level debugging level, but haven't come up with a lead. So far, the last message by WRF is related to  the generation of HALO stuff, but it doesn't necessarily mean the problem is there. I'll keep you posted.

PD: I'm using the new version of URBPARAM_LCZ.TBL you gave me.

**Comment by andreazonato** [comment]
@dargueso, after how much time it stops? before or after some time-step?

**Comment by dargueso** [comment]
@matthiasdemuzere, don't worry. I'll try to figure out what's going on, perhaps with @andreazonato's help.
@andreazonato, before any time-step, so the model itself doesn't actually get to run. It dies in the initialisation. I'm going to try run it locally, just to make sure it is not an issue with the compilation in the supercomputer, since I had to recompile it using your version.

**Comment by andreazonato** [comment]
did you set num_land_cat=41 in the namelist? Should be a problem of this kind...

**Comment by dargueso** [comment]
Yep, otherwise the model complains with that message. Let see how it goes with the serial version locally.

**Comment by dargueso** [comment]
Running in serial I get this error:
d02 2015-12-22_00:00:00 in NOAH DRV
-------------- FATAL CALLED ---------------
FATAL CALLED FROM FILE:  <stdin>  LINE:     774
ZDC + Z0C + 2m is larger than the 1st WRF level - Stop in subroutine urban - change ZDC and Z0C

I asume is related to urban LCZ parameters, but not sure. Can I get your help here?

**Comment by matthiasdemuzere** [comment]
without knowing WRF. Seems the few high buildings in our domain are intersecting with a low lowest WRF level (in the vertical)? So perhaps the height of the levels should be re-defined?

Just my two cents, @andreazonato will be better positioned to clarify this ....

**Comment by andreazonato** [comment]
I think it can be something like this. Actually I'm not an expert of the
SLUCM, i used to use BEP+BEM.
While BEP+BEM levels can be "subgrid" within WRF, the first WRF level
should always include all the city. So it can be as you say @Matthias
Demuzere ***@***.***> !

Il giorno mer 19 mag 2021 alle ore 16:52 Matthias Demuzere <
***@***.***> ha scritto:

**Comment by matthiasdemuzere** [comment]
> I get a bit lost with this level of detail...Should we redefine the URBPARAM then?

This solution would of course go against the main aim of our exercise, that is to address the added value of better and more explicit parameters.

How is the height of the model layers set in WRF? And what is the lowest height? Is this information also embedded into the boundary conditions, and are thus already in the parent domain?

What we could to:
- re-make a test geom_em file, with indeed lower artificial building height values. Just to see whether this works ...
- in case it is impossible to change the heights of the model layers, perhaps we should move to BEP+BEM after all? I realize this is then again different from the Li et al study, but since we introduced so many changes already, perhaps that does not matter anymore?

What do you both think @andreazonato @dargueso ?

**Comment by dargueso** [comment]
Yes, I agree with you. I was just wondering if there are ways of modifying it but keeping the city heterogeneity. Thinking about that, I realised it doesn't make much sense. 

The model layers in WRF can be defined automatically by the model or you can explicit them. But pushing the first level above the tallest building may be too high depending on the city. The model layers can be defined through the hybrid coordinates, so I will have to do some trial an error until it works. I can try and see if it needs a big change or just a small one. This should be done at the metgrid.exe stage, when vertical levels are defined. The geo_em should be fine. 

If that doesn't work, we can explore the BEP+BEM alternative. I must say that I have no experience at all running it and that we would have to throw away exp B results so far (20%).

**Comment by matthiasdemuzere** [comment]
Ha, yes, you are right ... If we change the scheme in Exp. C, then Exp. B becomes useless. That is of course also not good.

Ok, then ideally we can change the model levels a bit?
Yet to make sure that my bug fix suggestion works, before you start doing your work, should I not first make a new geom_em with lower building heights?

Any idea what the height of the lowest layer currently is?

The max building height in our domain is currently 50 m + 5 m std ...

**Comment by dargueso** [comment]
The first WRF level is roughly at 26 m AGL. We would have to double that, which will have some effects on the results...Shall I go with that option? 
The second level is at ~85m

**Comment by dargueso** [comment]
Models are quite sensitive to the location of the first level, so I would personally avoid that option. @andreazonato, how involved is switching from SLUCM to BEP+BEM? Do we need to make too many additional steps or is it straightforward by just changing namelist options?

**Comment by matthiasdemuzere** [comment]
That is low indeed. But was probably never an issue before, as - if I am not mistaken - the default value used in WRF is 7.5m for urban pixels?

Not sure what the best option would be here. If we change the lowest level, results for Exp. C. might in any case be different with Exp. C. because of this? 

Should I first make a boundary file with max BH of e.g. 20m, to see whether this is really causing the issue?

**Comment by dargueso** [comment]
Whatever you prefer, I can also push the lowest level up and see how it goes.

**Comment by andreazonato** [comment]
Dear all, sorry, pretty busy today, I'll have a look to your replies
tomorrow.

Anyway, smoothing a bit the height of tallest buildings in my opinion will
maintain the heterogeneity the same,without increasing the first vertical
level. It will affect just the cells with higher buildings. Switching to
bep+bem means simply to change the namelist. And I'm more used to bep bem
so I can help more. You should try!

Il gio 20 mag 2021, 10:51 Daniel Argüeso ***@***.***> ha
scritto:

**Comment by dargueso** [comment]
I've tried using BEP+BEM and it works, but it requires additional changes in the PBL scheme and Surface Layer Scheme. I imagine this is because BEP+BEM cannot work with a non-local scheme, and I would assume this may also partly explain why BEP+BEM works with buildings above the first level, as opposed to SLUCM.

Anyway, this is not a feasible option then, because we cannot compare with the natural one. So moving back to SLUCM, we can reduce the size of the tallest buildings or move the first level up.
I'm trying the second option, just to check it works. 

@matthiasdemuzere, if you want, you can smooth the buildings. That would mean that the we can run the model using the exact same configuration as Exp A and B.

If we decide for moving the first level up, we should keep that in mind when analysing the results, because differences can arise from that.

**Comment by andreazonato** [comment]
BEP+BEM, in the new version, can now work with YSU (if you are using it). you should simply download the latest WRF version, and there there is the coupling. In my opinion, smoothing a bit the highest buildings can be an option, as trying the BEP+BEM in the new WRF version

**Comment by dargueso** [comment]
I'm using v4.2.2. I think I'm using the modified version in the repository that has your changes to use WUDAPT. I've seen that v4.3 was released 10 days ago, which now allows BEP+BEM with YSU, but are your changes included there too?

**Comment by andreazonato** [comment]
Yes, my changes are included too! You should simply add @matthiasc2s 's URBPARM_LCZ.TBL plus a couple of lines that i'll send you, and that'all you need to modify

**Comment by dargueso** [comment]
OK, lets go with that option then. Send me the lines and I'll get v4.3 compiled locally and in the supercomputer.

**Comment by andreazonato** [comment]
#
# GR_FLAG: 1 to switch on green roof model (0-1)
#      (sf_urban_physics=3)
#

GR_FLAG:0

#
# GR_TYPE: 1 for GRASS, 2 for SEDUM vegetation on the green roof
#      (sf_urban_physics=3)
#

GR_TYPE: 2

#
# GR_FRAC_ROOF: fraction of green roof over the roof (0:1)
#      (sf_urban_physics=3)
#

GR_FRAC_ROOF:0,0,0,0,0,0,0,0,0,0,0

#
# HSEQUIP:  Diurnal  profile of sprinkler irrigation for the green roof
#      (sf_urban_physics=3)
#

IRHO:0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1


#
# PV_FRAC_ROOF: fraction of photovoltaic panels over the roof (0:1)
#      (sf_urban_physics=3)
#

PV_FRAC_ROOF: 0,0,0,0,0,0,0,0,0,0,0

These are the lines you should add to Matthias' URBPARM_LCZ.TBL!

**Comment by dargueso** [comment]
Thanks! I'll give it a go

**Comment by matthiasdemuzere** [comment]
Good work guys.

The additional lines that were sent are also [here](https://github.com/wrf-model/WRF/blob/f4f83984357b021062e6ff94a21fe7aac3e9937b/run/URBPARM_LCZ.TBL), as I pushed this new default table to WRF's new v4.3.1 release. So it should be present in an upcoming version.

Just for you to know ...

**Comment by dargueso** [comment]
The option of moving the 1st level up works. I changed the 1st level in the model coordinates and resulted in the first level at around 45m (I aimed at 55, but I was just roughly selecting the model levels). Anyway, it seems to work with that modification. 
Shall we move forward with this option and keep what has been run for exp B (~20%) or move to BEP+BEM? I don't have a strong opinion on this.

We might run out of CPU time before completing both, but I already put in another application and allowed for some extra runs of this side-project.

**Comment by andreazonato** [comment]
The point is this, in my opinion:

if you move up the 1st level, then you should re-run exp A) and B), since
conditions changes in the numerical computation. For this reason, I would
remain with the possibility of:

1) smooth tallest buildings
2) run it with BEP+ BEM.



Il giorno ven 21 mag 2021 alle ore 12:29 Daniel Argüeso <
***@***.***> ha scritto:

**Comment by dargueso** [comment]
Agree. I'm compiling v4.3 and will see how it goes with BEP+BEM. I will thus rerun expB.

**Comment by matthiasdemuzere** [comment]
I agree with this path taking. The most clean one, in which we'll not have to talk ourselves out of certain decisions we took.
And ... beneficial because Andrea is more experienced with BEP+BEM.

@dargueso: also thanks for pro-actively requesting additional CPU time.

@dargueso: can you hold on a little bit longer with submitting Exp. C. @albertomartilli found some strange behaviour of my building height distribution implementation for the Zaragoza case, and I'd like to have a closer look at this first. I was hoping to do this today, yet might be only able to do this in towards the end of next week?

**Comment by dargueso** [comment]
Ok, we'll take that path!

I'll hold exp C until you tell me, @matthiasdemuzere. I'll be busy compiling and running exp B.

**Comment by albertomartilli** [comment]
Hi All,
I read the whole chain..very iteresting. Yes, SLUCM cannot have the first level below building height. Mind also that the T2 (2m temperature's output of WRF) with SLUCM is in reality 2m above roof level - it is derived with a log-law using the total heat flux, including the one from roofs).

May I know what Exp. A, B, and C are about?

**Comment by matthiasdemuzere** [comment]
Hi @albertomartilli,

Thanks for the info. Now that we are moving to BEP/BEM, this item is solved then.

With respect to the experiments:

The discussion of [Li et al (2020)](http://doi.org/10.1029/2020GL088758) on urban induced precip stated that future work should include a more heterogeneous urban description. So I contacted Daniel to re-do this experiment, using LCZ information. More details on this can be found in the [README](https://github.com/matthiasdemuzere/wrf-lcz-KL/), yet it comes down to:
- Exp A.: removing urban with natural land cover (as in Li et al.)
- Exp. B: LCZ urban extent, no urban parameters (as in Li et al.)
- Exp. C: full LCZ information (new, to test the added value of LCZ params)

So besides this research question, we also hope to get this tool out ...

FYI: I opened a new issue [on the bug in HI_URB2D](https://github.com/matthiasdemuzere/wrf-lcz-KL/issues/18). I will report progress on this there.
